src/pineko/kfactor.py

```diff
@@ -209,168 +168,129 @@ def create_singlegridonly(
+    """Cast the centrals k-factors to the correct length.
+
+    Apply a normalization according to the length compared to the number of bins of the grid.
+
+    Parameters
+    ----------
+    grid: pineappl.grid.Grid
+        grid
+    centrals_kfactor: list
+        list of kfactor for each point
+    """
+    if grid.bins() == len(central_kfactors):
+        rich.print("The number of bins match the length of the k-factor.")
+        return central_kfactors
+    if grid.bins() < len(central_kfactors):
         rich.print(
-            "[yellow] The number of bins is more than the lenght of the k-factor."
+            "[yellow] The number of bins is less than the length of the k-factor."
         )
-
-        # This is the last case in which grid.bins() > len(centrals_kfactor)
-
-        # Note that sometimes there are more bins in the grid than in the cfactor file -
-        # this is not a problem because in those cases either all cfactor values are the
-        # same (thus there is no doubt about whether we have the correct one) or the
-        # non-exisiting cfactors would be multiplied by bins corresponding to all '0' in the
-        # grid.
-        # Let's check if we are in the first or second case
-        if len(np.unique(centrals_kfactor)) == 1:
-            # In this case I just need to add more elements to the kfactor
-            for _num in range(pigrid.bins()):
-                centrals_kfactor_filtered = np.append(
-                    centrals_kfactor_filtered, centrals_kfactor[0]
-                )
-        else:
-            # In this case this means that the missing entries will multiply zero subgrids so we can just add 0s
-            for _num in range(pigrid.bins()):
-                centrals_kfactor_filtered = np.append(centrals_kfactor_filtered, 0.0)
-    return centrals_kfactor_filtered
+        if not len(np.unique(central_kfactors)) == 1:
+            # This case is actually wrong.
+            raise ValueError("k-factor contains too many different values.")
+        return central_kfactors
+
+    rich.print("[yellow] The number of bins is more than the length of the k-factor.")
+
+    # This is the last case in which grid.bins() > len(centrals_kfactor)
+    # Note that sometimes there are more bins in the grid than in the kfactor file -
+    # this is not a problem because in those cases either all kfactor values are the
+    # same (thus there is no doubt about whether we have the correct one) or the
+    # non-exisiting kfactor would be multiplied by bins corresponding to all '0' in the
+    # grid.
+    # Let's check if we are in the first or second case
+    if len(np.unique(central_kfactors)) == 1:
+        # In this case I just need to add more elements to the kfactor
+        return np.full(grid.bins(), central_kfactors[0])
+    # In this case this means that the missing entries will
+    # multiply zero subgrids so we can just add 0s
+    return np.full(grid.bins(), 0)
 
 
 def compute_k_factor_grid(
```

Comment:
```suggestion
def apply_to_dataset(
```

Reply:
done in b62249b9271626b2ab78f71235e3939676758df5